### PR TITLE
Unify Kotlin and JavaScript builds

### PR DESCRIPTION
**Pull request by xian**
* Switched to Kotlin DSL for gradle build files.
* Updated to Kotlin 1.3.70-eap-274.
* Unified webpack for Kotlin and JSX code.
* Reworked FakeClientDevice/HostedWebApp.
* UI app properly unsubscribes from pubsub channels when shutting down.
* Specify JS npm dependencies in the gradle build.

### CODE REVIEW

**Comment by xian** [comment]
@hassoncs and @r3dcrosse — this is a pretty big PR, sorry. Key things are that the JS/JSX build now happens from within the Gradle build, and both JS and Kotlin code can call more naturally into each other. Lemme know if you see anything scary!

File: src/jsMain/kotlin/baaahs/WebUi.kt [hunk 1]
```diff
@@ -1,12 +1,30 @@
...
+
+    override fun render(): ReactElement {
+        val webUiClientLink = network.link()
+        val pubSub = PubSub.Client(webUiClientLink, pinkyAddress, Ports.PINKY_UI_TCP).apply {
             install(gadgetModule)
         }
-}
\ No newline at end of file
+
+        if (1 + 1 == 3) {
```
Comment (r3dcrosse) [inline src/jsMain/kotlin/baaahs/WebUi.kt]: Wait, what? Is this a thing we need to worry about?

File: src/jsMain/resources/ui-index.html [hunk 2]
```diff
@@ -14,30 +14,12 @@ <h3>Sparkle Motion UI <small><i>FUCK YEAH</i></small></h3>
...
-
-    require(['react_app.js'], function (reactApp) {
-      let network = new baaahs.net.BrowserNetwork();
-      let pinkyAddress = new baaahs.net.BrowserNetwork.BrowserAddress(myUrl());
-      let pubSub = baaahs.WebUi.createPubSubClient(network, pinkyAddress);
-      let webApp = document.createUiApp(pubSub);
-      webApp.render(document.body);
-    });
-  });
+  document.sparklemotionMode = "UI";
```
Comment (r3dcrosse) [inline src/jsMain/resources/ui-index.html]: Yay, this is much better!

File: src/jvmMain/kotlin/baaahs/jvmImpl.kt [hunk 3]
```diff
@@ -54,13 +55,13 @@ class RealFs(private val basePath: Path) : Fs {
     }
 }
 
-actual fun logMessage(level: String, message: String, exception: Throwable?) {
+actual fun log(id: String, level: String, message: String, exception: Throwable?) {
+    val logger = LoggerFactory.getLogger(id)
```
Comment (r3dcrosse) [inline src/jvmMain/kotlin/baaahs/jvmImpl.kt]: Does this log to a log file now? Or is it still logging in the browser console?

File: src/jsMain/kotlin/baaahs/WebUi.kt (hunk 1 above)
Comment (xian) [reply src/jsMain/kotlin/baaahs/WebUi.kt]: I'm sure there's a better way to do it, I was just in a hurry. :-)

File: src/jvmMain/kotlin/baaahs/jvmImpl.kt (hunk 3 above)
Comment (xian) [reply src/jvmMain/kotlin/baaahs/jvmImpl.kt]: Logs in the browser still go to console. The jvm Pinky now logs through the normal Byzantine maze of Java logging frameworks.

**Pull request merged**